### PR TITLE
Add perf event support

### PR DESCRIPTION
**Pull request by caizixian**
This PR supports measuring perf event for work packets.

The added code is guarded behind a feature `perf` due to an upstream bug in bindgen.
https://github.com/rust-lang/rust-bindgen/issues/1538
https://github.com/rust-lang/rust/issues/59154

### CODE REVIEW

**Comment by caizixian** [comment]
https://github.com/rust-lang/cargo/issues/2524
https://github.com/rust-lang/cargo/issues/7914
https://github.com/rust-lang/cargo/issues/1197

**Comment by k-sareen** [comment]
So I don't know if this is a bug with the `pfm` library or this PR, but on an Intel machine (i9 9900K), the combination of:
```
MMTK_PERF_EVENTS="PERF_COUNT_HW_CACHE_MISSES,0,-1;PERF_COUNT_HW_BRANCH_MISSES,0,-1;PERF_COUNT_HW_CACHE_L1I:MISS,0,-1;PERF_COUNT_HW_CACHE_L1D:MISS,0,-1;PERF_COUNT_HW_INSTRUCTIONS,0,-1;PERF_COUNT_HW_CACHE_DTLB:MISS,0,-1"
```
consistently fails the assertion at [scheduler/work_counter.rs:237](https://github.com/mmtk/mmtk-core/pull/315/files#diff-9f37824a4dd5939d512658d8f6f9df1914479935eb54d77a40b34163c1657979R237):

```
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `32645`,
 right: `0`', /home/kunals/git/mmtk-core/src/scheduler/work_counter.rs:237:13
```

I tried various combinations of the above perf events and found only one more that failed:
```
MMTK_PERF_EVENTS="PERF_COUNT_HW_CACHE_MISSES,0,-1;PERF_COUNT_HW_BRANCH_MISSES,0,-1;PERF_COUNT_HW_CACHE_L1I:MISS,0,-1;PERF_COUNT_HW_CACHE_L1D:MISS,0,-1;PERF_COUNT_HW_CACHE_DTLB:MISS,0,-1"
```
and every other combination I tried works as expected.

It's weird because there are a lot of combinations that do work on the Intel machine as well as all the combinations work on a AMD (Ryzen 9 3900X) machine.

Any ideas what's up?

File: src/util/statistics/stats.rs [hunk 1]
```diff
@@ -55,9 +60,18 @@ impl Stats {
...
+        // scope
+        #[cfg(feature = "perf")]
+        let mut perfmon: Perfmon = Default::default();
+        #[cfg(feature = "perf")]
+        perfmon.initialize().expect("Perfmon failed to initialize");
         Stats {
             gc_count: AtomicUsize::new(0),
             total_time: t.clone(),
+            #[cfg(feature = "perf")]
+            perfmon,
```
Comment (qinsoon) [inline src/util/statistics/stats.rs:74]: You can do this:
```suggestion
            perfmon: {
              let mut perfmon: Perfmon = Default::default();
              perfmon.initialize().expect("Perfmon failed to initialize");
              perfmon
            }
```

File: src/memory_manager.rs [hunk 2]
```diff
@@ -169,6 +169,14 @@ pub fn start_worker<VM: VMBinding>(
...
 pub fn enable_collection<VM: VMBinding>(mmtk: &'static MMTK<VM>, tls: VMThread) {
+    // initialize a list of perf events for work thread initialization
+    #[cfg(feature = "perf")]
+    {
+        use crate::scheduler::stat::PERF_EVENTS;
+        use crate::scheduler::work_counter::parse_perf_events;
+        let mut w = PERF_EVENTS.write().unwrap();
+        *w = parse_perf_events(&mmtk.options.perf_events)
+    }
     mmtk.scheduler.initialize(mmtk.options.threads, mmtk, tls);
```
Comment (qinsoon) [inline src/memory_manager.rs:180]: You can change the first parameter in `initialize()` to be `Options`, and do the initialization in `scheduler.initialize()`.

File: src/scheduler/mod.rs [hunk 3]
```diff
@@ -14,8 +14,8 @@ pub(crate) use scheduler::CoordinatorMessage;
 pub(crate) use scheduler::MMTkScheduler;
 pub(self) use scheduler::Scheduler;
 
-mod stat;
-mod work_counter;
+pub(crate) mod stat;
```
Comment (qinsoon) [inline src/scheduler/mod.rs:17]: This is unnecessary if the initialization is within the `schedule` mod.

File: src/scheduler/stat.rs [hunk 4]
```diff
@@ -3,6 +3,13 @@ use super::work_counter::{WorkCounter, WorkCounterBase, WorkDuration};
 use std::any::TypeId;
 use std::collections::HashMap;
 use std::sync::atomic::{AtomicBool, Ordering};
+#[cfg(feature = "perf")]
+use std::sync::RwLock;
+
+#[cfg(feature = "perf")]
+lazy_static! {
+    pub static ref PERF_EVENTS: RwLock<Vec<(String, i32, i32)>> = RwLock::new(vec![]);
```
Comment (qinsoon) [inline src/scheduler/stat.rs:11]: It seems unnecessary to have this global to temporarily store the parsed events. You could simply have a `PerfEventOptions(Vec<PerfEventConfig>)` in `Options`, and implement `FromStr` for the type. When we process options, the perf event option will be parsed and saved. When creating `WorkerLocalStats`, you just need a reference to `Options` or `MMTK` to access the options.

Comment (qinsoon) [reply src/scheduler/stat.rs:11]: `validate_perf_events` and `parse_perf_events` can be all put to the new `PerfEventOptions` type.

**Review by qinsoon — COMMENTED** [review]
I suggest 1. creating a new type for `PerfEventOptions`, and use that in the `Options` rather than using a `String`, and 2. getting rid of the `PERF_EVENTS` global.

File: Cargo.toml [hunk 5]
```diff
@@ -41,6 +42,10 @@ rand = "0.7.3"
 [features]
 default = []
 
+# This feature is only supported on x86-64 for now
+# Not included below so that CI won't fail
+perf = ["pfm"]
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply Cargo.toml:47]: And possibly in `ci-style.sh` as well.

File: src/util/statistics/stats.rs [hunk 6]
```diff
@@ -37,6 +40,8 @@ impl SharedStats {
 pub struct Stats {
     gc_count: AtomicUsize,
     total_time: Arc<Mutex<Timer>>,
+    #[cfg(feature = "perf")]
+    perfmon: Perfmon,
```
[the remark this answers is not in the document]
Comment (caizixian) [reply src/util/statistics/stats.rs:46]: Comment added

**Review by qinsoon — APPROVED** [review]

**Pull request merged**